### PR TITLE
geo/geomfn: implement ST_MakePolygon({geometry,_geometry})

### LINKED ISSUE #49305: geo/geomfn: implement ST_MakePolygon({geometry,_geometry})
Implement `ST_MakePolygon` on arguments {geometry,_geometry}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_MakePolygon.html).

_Observers: Please react to this issue if you need this functionality._

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-06-30T05:29:57Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by hiepd**
Closes #49305

* Implement ST_MakePolygon

Release note (sql change): Implement ST_MakePolygon functionality for geometry operator

|                                                                                                                                                                                                                                                                | Postgres                                                                                                 | CockroachDB                                                                                                             |
|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------|
| SELECT ST_MakePolygon('MULTIPOINT(0 0, 1 1)');                                                                                                                                                                                                                 | ERROR: Shell is not a line                                                                               | ERROR: st_makepolygon(): argument must be LINESTRING geometries                                                         |
| SELECT ST_MakePolygon('abc');                                                                                                                                                                                                                                  | HINT: "ab" <-- parse error at position 2 within geometry                                                 | ERROR: unknown signature: st_makepolygon(string)                                                                        |
| SELECT ST_MakePolygon( ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 80)'), ARRAY[ ST_GeomFromText('MULTIPOINT(50 70, 70 70, 70 50, 50 50, 50 70)') ]);                                                                                           | ERROR: Hole 0 is not a line                                                                              | ERROR: st_makepolygon(): argument must be LINESTRING geometries                                                         |
| SELECT ST_MakePolygon( ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 80)'), ARRAY['abc']);                                                                                                                                                        | ERROR: parse error - invalid geometry HINT: "ab" <-- parse error at position 2 within geometry           | ERROR: st_makepolygon(): argument must be LINESTRING geometries                                                         |
| SELECT ST_MakePolygon( ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 80)', 4326), ARRAY[ ST_GeomFromText('LINESTRING(50 70, 70 70, 70 50, 50 50, 50 70)') ]);                                                                                     | ERROR: lwpoly_from_lwlines: mixed SRIDs in input lines                                                   | ERROR: st_makepolygon(): mixed SRIDs are not allowed                                                                    |
| SELECT ST_MakePolygon( ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 80)', 4326), ARRAY[ ST_GeomFromText('LINESTRING(50 70, 70 70, 70 50, 50 50, 50 70)', 3857) ]);                                                                               | ERROR: lwpoly_from_lwlines: mixed SRIDs in input lines                                                   | ERROR: st_makepolygon(): mixed SRIDs are not allowed                                                                    |
| SELECT ST_MakePolygon( ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 80)', 4326), ARRAY[ ST_GeomFromText('LINESTRING(50 70, 70 70, 70 50, 50 50, 50 70)', 4326), ST_GeomFromText('LINESTRING(60 60, 75 60, 75 45, 60 45, 60 60)', 3857) ]);       | ERROR: lwpoly_from_lwlines: mixed SRIDs in input lines                                                   | ERROR: st_makepolygon(): mixed SRIDs are not allowed                                                                    |
| SELECT ST_MakePolygon(ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 70)'));                                                                                                                                                                       | ERROR:  lwpoly_from_lwlines: shell must be closed                                                        | ERROR: st_makepolygon(): shell must be closed                                                                           |
| SELECT ST_MakePolygon(ST_GeomFromText('LINESTRING(40 80, 80 80, 40 80)'));                                                                                                                                                                                     | ERROR:  lwpoly_from_lwlines: shell must have at least 4 points                                           | ERROR: st_makepolygon(): shell must have at least 4 points                                                              |
| SELECT ST_MakePolygon(     ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 80)'),     ARRAY[       ST_GeomFromText('LINESTRING(50 70, 70 70, 70 50, 50 50, 50 60)'),       ST_GeomFromText('LINESTRING(60 60, 75 60, 75 45, 60 45, 60 60)')     ]); | ERROR:  lwpoly_from_lwlines: holes must be closed                                                        | ERROR: st_makepolygon(): holes must be closed                                                                           |
| SELECT ST_MakePolygon(     ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 80)'),     ARRAY[       ST_GeomFromText('LINESTRING(50 70, 70 70, 50 70)')     ]);                                                                                       | ERROR:  lwpoly_from_lwlines: holes must have at least 4 points                                           | ERROR: st_makepolygon(): holes must have at least 4 points                                                              |
| SELECT ST_AsEWKT(ST_MakePolygon( ST_GeomFromText('LINESTRING(75 29,77 29,77 29, 75 29)')));                                                                                                                                                                    | POLYGON((75 29,77 29,77 29,75 29))                                                                       | POLYGON ((75 29, 77 29, 77 29, 75 29))                                                                                  |
| SELECT ST_AsEWKT(ST_MakePolygon( ST_GeomFromText('LINESTRING(75 29,77 29,77 29, 75 29)', 4326) ));                                                                                                                                                             | SRID=4326;POLYGON((75 29,77 29,77 29,75 29))                                                             | SRID=4326;POLYGON ((75 29, 77 29, 77 29, 75 29))                                                                        |
| SELECT ST_AsEWKT(ST_MakePolygon( ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 80)'), ARRAY[ ST_GeomFromText('LINESTRING(50 70, 70 70, 70 50, 50 50, 50 70)') ]));                                                                                | POLYGON((40 80,80 80,80 40,40 40,40 80),(50 70,70 70,70 50,50 50,50 70))                                 | POLYGON ((40 80, 80 80, 80 40, 40 40, 40 80), (50 70, 70 70, 70 50, 50 50, 50 70))                                      |
| SELECT ST_AsEWKT(ST_MakePolygon( ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 80)'), ARRAY[ ST_GeomFromText('LINESTRING(50 70, 70 70, 70 50, 50 50, 50 70)'), ST_GeomFromText('LINESTRING(60 60, 75 60, 75 45, 60 45, 60 60)') ]));              | POLYGON((40 80,80 80,80 40,40 40,40 80),(50 70,70 70,70 50,50 50,50 70),(60 60,75 60,75 45,60 45,60 60)) | POLYGON ((40 80, 80 80, 80 40, 40 40, 40 80), (50 70, 70 70, 70 50, 50 50, 50 70), (60 60, 75 60, 75 45, 60 45, 60 60)) |
| SELECT ST_AsEWKT(ST_MakePolygon( ST_GeomFromText('LINESTRING(40 80, 80 80, 80 40, 40 40, 40 80)', 4326), ARRAY[ ST_GeomFromText('LINESTRING(50 70, 70 70, 70 50, 50 50, 50 70)', 4326) ]));                                                                    | SRID=4326;POLYGON((40 80,80 80,80 40,40 40,40 80),(50 70,70 70,70 50,50 50,50 70))                       | SRID=4326;POLYGON ((40 80, 80 80, 80 40, 40 40, 40 80), (50 70, 70 70, 70 50, 50 50, 50 70))                            |

## NOTES

Since Geometry from geo package does not support 3D, the following are ignored for now:
- create 3D polygon from 3D linestrings or linestrings with measures

### CODE REVIEW

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 1]
```diff
@@ -530,6 +530,47 @@ var geoBuiltins = map[string]builtinDefinition{
...
+					interior[i] = v.(*tree.DGeometry).Geometry
+				}
+				g, err := geomfn.MakePolygon(outer.Geometry, interior...)
+				if err != nil {
+					return nil, err
+				}
+				return tree.NewDGeometry(g), nil
+			},
+			Info: infoBuilder{
+				info: `Returns a new Polygon with the given outer linestring and interior (hole) linestrings.`,
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:569]: can you capitalise linestring as LineString?
same with the above.

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 2]
```diff
@@ -530,6 +530,47 @@ var geoBuiltins = map[string]builtinDefinition{
...
+				{"outer", types.Geometry},
+				{"interior", types.AnyArray},
+			},
+			ReturnType: tree.FixedReturnType(types.Geometry),
+			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+				outer := args[0].(*tree.DGeometry)
+				interiorArr := tree.MustBeDArray(args[1])
+				interior := make([]*geo.Geometry, len(interiorArr.Array))
+				for i, v := range interiorArr.Array {
+					interior[i] = v.(*tree.DGeometry).Geometry
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:560]: We should probably check the type assertion holds true lest we panic, i.e. `interiorRing, ok := v.(*tree.DGeometry)`  first. Add a test for this too :)

File: pkg/geo/geomfn/polygon.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// MakePolygon creates a Polygon geometry from linestring and optional inner linestrings.
+// Returns errors if geometries are not linestrings.
+func MakePolygon(outer *geo.Geometry, interior ...*geo.Geometry) (*geo.Geometry, error) {
+	outerGeomT, err := outer.AsGeomT()
+	if err != nil {
+		return nil, err
+	}
+	outerLineStr, ok := outerGeomT.(*geom.LineString)
+	if !ok {
+		return nil, errors.Newf("Argument must be LINESTRING geometries")
```
Comment (otan) [inline pkg/geo/geomfn/polygon.go:28]: first word of an error message should be lowercase, i.e. "argument must be"

File: pkg/geo/geomfn/polygon.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	}
+	coords := make([][]geom.Coord, len(interior)+1)
+	coords[0] = outerLineStr.Coords()
+	for i, ring := range interior {
+		ringGeomT, err := ring.AsGeomT()
```
Comment (otan) [inline pkg/geo/geomfn/polygon.go:33]: nit: name this interiorRing, and outerLineStr to "outerRing".

**Review by otan — COMMENTED** [review]
Thanks for submitting this! Just a couple of small nits.

File: pkg/geo/geomfn/polygon.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+			return nil, err
+		}
+		ringLineStr, ok := ringGeomT.(*geom.LineString)
+		if !ok {
+			return nil, errors.Newf("Argument must be LINESTRING geometries")
+		}
+		coords[i+1] = ringLineStr.Coords()
+	}
+
+	polygon, err := geom.NewPolygon(geom.XY).SetCoords(coords)
```
Comment (otan) [inline pkg/geo/geomfn/polygon.go:44]: does the SRID need to be set on the object?
can you add tests for setting srid on the outer LineString and the innterlinestring should match / SRID set in shell but not the rings / mismatching SRIDs set (and check behaviour matches PostGIS)?

File: pkg/geo/geomfn/polygon.go (hunk 4 above)
Comment (manhhiep92) [reply pkg/geo/geomfn/polygon.go]: Updated. Thanks!

File: pkg/geo/geomfn/polygon.go (hunk 3 above)
Comment (manhhiep92) [reply pkg/geo/geomfn/polygon.go]: Updated. Thanks!

File: pkg/sql/sem/builtins/geo_builtins.go (hunk 2 above)
Comment (manhhiep92) [reply pkg/sql/sem/builtins/geo_builtins.go]: Added a check for this. I also updated the PR description for Postgres and CockroachDB behaviors for `ST_MakePolygon`
Thanks!

File: pkg/geo/geomfn/polygon.go (hunk 5 above)
Comment (manhhiep92) [reply pkg/geo/geomfn/polygon.go]: Good catch! Thanks. Added logic to set SRID on the geometry, unit tests and also logic tests. 

I also added behaviors between Postgres and CRDB for ST_MakePolygon in the PR description as well.

**Review by otan — COMMENTED** [review]
Please also refactor your commit message to:

```
geo/geomfn: implement ST_MakePolygon({geometry,_geometry})

Closes #49305

Release note (sql change): Implement ST_MakePolygon functionality
for Geometry types.
```

Note title the important thing that the title is the first line of the commit message.

Thanks!

File: pkg/geo/geomfn/polygon.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+			return nil, errors.Newf("mixed SRIDs are not allowed")
+		}
+		coords[i+1] = interiorRing.Coords()
+	}
+
+	polygon, err := geom.NewPolygon(geom.XY).SetSRID(srid).SetCoords(coords)
+	if err != nil {
+		return nil, err
+	}
+	return geo.NewGeometryFromGeom(polygon)
```
Comment (otan) [inline pkg/geo/geomfn/polygon.go:52]: You also need to rebase this -- this should be `NewGeometryFromGeomT`.

File: pkg/geo/geomfn/polygon.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+		}
+		interiorRing, ok := interiorRingGeomT.(*geom.LineString)
+		if !ok {
+			return nil, errors.Newf("argument must be LINESTRING geometries")
+		}
+		if interiorRing.SRID() != srid {
+			return nil, errors.Newf("mixed SRIDs are not allowed")
+		}
+		coords[i+1] = interiorRing.Coords()
+	}
```
Comment (otan) [inline pkg/geo/geomfn/polygon.go:46]: one more minor detail: can you also add a test to make sure linestrings are closed with at least four points, e.g.

```
otan=# select st_makepolygon('LINESTRING(0 0, 1 0)');
ERROR:  lwpoly_from_lwlines: shell must have at least 4 points
otan=# select st_makepolygon('LINESTRING(0 0, 1 0, 1 1, 0.1 0.1)');
ERROR:  lwpoly_from_lwlines: shell must be closed
```

Comment (manhhiep92) [reply pkg/geo/geomfn/polygon.go:59]: Added a check for number of points and closure. Thanks! Not sure if this is only Postgis behavior or `geom` should also implement this check as well.

Comment (otan) [reply pkg/geo/geomfn/polygon.go:59]: Probably, but we'll make do with this for now :)

**Review by otan — APPROVED** [review]
Thanks!

File: pkg/geo/geomfn/polygon.go (hunk 7 above)
Comment (otan) [reply pkg/geo/geomfn/polygon.go:59]: Filed #51074.

**Comment by Amruta-Ranade** [comment]
@hiepd Thank you for contributing to CockroachDB this year. As a token of our appreciation, we would like to send you a gift. Please DM me on our [community Slack](https://cockroa.ch/slack) @Amruta so I can send you a link. (If you don’t want a gift, we also have a charitable contribution choice.) All orders need to be in by December 13, so please contact me as soon as possible :)

**Pull request merged**